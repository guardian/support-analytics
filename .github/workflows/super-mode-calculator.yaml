name: Upload super-mode-calculator to riff raff
on:
  pull_request:
    branches:
      - main
  push:
    branches:
      - main
jobs:
  super_mode_calculator_build:
    if: github.repository_owner == 'guardian'
    # Required by actions-riff-raff
    permissions:
      id-token: write
      contents: read

    name: Build super-mode-calculator
    runs-on: ubuntu-latest

    steps:
      - name: Env
        run: env

      - name: Checkout repo
        uses: actions/checkout@v3

      - name: Configure AWS credentials # needed by actions-riff-raff
        uses: aws-actions/configure-aws-credentials@v3
        with:
          role-to-assume: ${{ secrets.GU_RIFF_RAFF_ROLE_ARN }}
          aws-region: eu-west-1

      - name: Setup Node
        uses: actions/setup-node@v3
        with:
          cache: 'yarn'
          cache-dependency-path: ./super-mode-calculator/yarn.lock
          node-version-file: super-mode-calculator/.nvmrc

      - name: Yarn
        run: |
          yarn install
          yarn lint
          yarn run build
        working-directory: ./super-mode-calculator

      - name: Copy files to Riff Raff package
        run: cp package.json riff-raff.yaml target
        working-directory: ./super-mode-calculator

      - name: Yarn install in package
        run: yarn install --production
        working-directory: ./super-mode-calculator/target

      - name: Zip target directory contents (quietly)
        run: zip -qr super-mode-calculator.zip ./*
        working-directory: ./super-mode-calculator/target

      - uses: guardian/actions-riff-raff@v2
        with:
          projectName: Support::super-mode-calculator
          configPath: ./super-mode-calculator/riff-raff.yaml
          buildNumberOffset: 250
          contentDirectories: |
            super-mode-calculator:
              - ./super-mode-calculator/target/super-mode-calculator.zip
            super-mode-calculator-cloudformation:
              - ./super-mode-calculator/cfn.yaml
