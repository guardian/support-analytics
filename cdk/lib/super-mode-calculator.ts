import {GuScheduledLambda} from "@guardian/cdk";
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import {GuStack, GuStringParameter} from '@guardian/cdk/lib/constructs/core';
import {type App, Duration, RemovalPolicy} from 'aws-cdk-lib';
import {AttributeType, BillingMode, Table} from "aws-cdk-lib/aws-dynamodb";
import { Schedule } from 'aws-cdk-lib/aws-events';
import { PolicyStatement, Role, ServicePrincipal} from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

const appName = 'super-mode-calculator-app';
export class SuperModeCalculator extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const scheduleState = new GuStringParameter(this, 'ScheduleState', {
			description: 'The state of the schedule',
			default: 'ENABLED',
			allowedValues: ['ENABLED', 'DISABLED'],
		});

		const scheduleRules =
			scheduleState.valueAsString === 'ENABLED'
				? [
					{
						schedule: Schedule.rate(Duration.minutes(60)),
					},
				]
				: [];

		const superModeCalculatorTable = new Table(this, 'super-mode-calculator-table', {
			tableName: `super-mode-${this.stage}`,
			removalPolicy: RemovalPolicy.RETAIN,
			billingMode: BillingMode.PROVISIONED,
			partitionKey: {
				name: 'id',
				type: AttributeType.STRING,
			},
			sortKey: {
				name: 'startTimestamp',
				type: AttributeType.STRING,
			},
			pointInTimeRecovery: this.stage === 'PROD',
		});



		const role = new Role(this, 'query-lambda-role', {
			// Set the name of the role rather than using an autogenerated name.
			// This is because if the ARN is too long then it breaks the authentication request to GCP
			roleName: `${appName}-${this.stage}`,
			assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
		});
		role.addToPolicy(
			// Logging permissions
			new PolicyStatement({
				actions: [
					'logs:CreateLogGroup',
					'logs:CreateLogStream',
					'logs:PutLogEvents',
				],
				resources: ['*'],
			}),
		);
		role.addToPolicy(
			// Permission to read config from Parameter Store
			new PolicyStatement({
				actions: ['ssm:GetParameter'],
				resources: [
					`arn:aws:ssm:${this.region}:${this.account}:parameter/super-mode-calculator/${this.stage}/gcp-wif-credentials-config`,
				],
			}),
		);
		role.addToPolicy(
				new PolicyStatement({
					actions: ['dynamodb:BatchWriteItem'],
					resources: [superModeCalculatorTable.tableArn],
				}),
		);

		new GuScheduledLambda(this, 'SuperModeCalculator', {
			app: appName,
			functionName: `${appName}-${this.stage}`,
			runtime: Runtime.NODEJS_20_X,
			handler: 'super-mode-calculator/src/lambdas/lambda.handler',
			fileName: `${appName}.zip`,
			rules: scheduleRules,
			role,
			monitoringConfiguration:
				this.stage === 'PROD'
					? {
						toleratedErrorPercentage: 0,
						snsTopicName: 'alarms-handler-topic-PROD',
					}
					: { noMonitoring: true },
		});
	}
}
