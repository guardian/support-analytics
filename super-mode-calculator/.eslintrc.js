module.exports = {
  extends: ["@guardian/eslint-config-typescript"],
};
