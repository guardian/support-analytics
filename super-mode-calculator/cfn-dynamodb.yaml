AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Description: Super mode dynamodb table

Parameters:
  Stage:
    Description: Stage name
    Type: String
    AllowedValues:
      - CODE
      - PROD

Resources:
  SuperMode:
    Type: AWS::DynamoDB::Table
    Properties:
      TableName: !Sub super-mode-${Stage}
      AttributeDefinitions:
          # id is "<region>_<b64-encoded url>", e.g. "US_https://theguardian.com/..."
        - AttributeName: id
          AttributeType: S
        - AttributeName: startTimestamp
          AttributeType: S
        - AttributeName: endTimestamp
          AttributeType: S
        - AttributeName: endDate
          AttributeType: S
      KeySchema:
        - AttributeName: id
          KeyType: HASH
        - AttributeName: startTimestamp
          KeyType: RANGE
      ProvisionedThroughput:
        ReadCapacityUnits: 5
        WriteCapacityUnits: 5
      GlobalSecondaryIndexes:
        - IndexName: end
          KeySchema:
            - AttributeName: endDate
              KeyType: HASH
            - AttributeName: endTimestamp
              KeyType: RANGE
          Projection:
            ProjectionType: ALL
          ProvisionedThroughput:
            ReadCapacityUnits: 5
            WriteCapacityUnits: 5
