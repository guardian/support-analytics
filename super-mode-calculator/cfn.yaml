AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Description: Super mode calculator state machine

Parameters:
  Stack:
    Description: Stack name
    Type: String
    Default: support
  App:
    Description: Application name
    Type: String
    Default: super-mode-calculator
  Stage:
    Description: Stage name
    Type: String
    AllowedValues:
      - CODE
      - PROD
  DeployBucket:
    Description: Bucket to copy files to
    Type: String
    Default: membership-dist
  AthenaOutputBucket:
    Description: Name of the bucket to output athena query results to
    Type: String
    Default: gu-support-analytics
  AcquisitionEventsBucket:
    Description: Bucket containing the acquisition events data to be queried
    Type: String
    Default: acquisition-events
  SchemaName:
    Description: The schema name for the query
    Type: String
    Default: acquisition
  CronExpression:
    Description: Cron expression for scheduling the step functions
    Type: String
    Default: 00 * * * ? *
  ScheduleState:
    Description: The state of the schedule
    Type: String
    Default: ENABLED
    AllowedValues:
      - ENABLED
      - DISABLED

Conditions:
  IsProd: !Equals [!Ref "Stage", "PROD"]

Resources:
  PartitionLambda:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: !Sub super-mode-calculator-partition-lambda-${Stage}
      Description: Creates Athena paritions for querying epic views and acquisitions
      Runtime: nodejs14.x
      Handler: lambdas/partition/lambda.handler
      MemorySize: 128
      Timeout: 300
      Environment:
        Variables:
          Stage: !Ref Stage
          AthenaOutputBucket: !Ref AthenaOutputBucket
          SchemaName: !Ref SchemaName
      CodeUri:
        Bucket: !Ref DeployBucket
        Key: !Sub ${Stack}/${Stage}/${App}/${App}.zip
      Policies:
        - AWSLambdaBasicExecutionRole
        - AmazonAthenaFullAccess
        - Statement:
            Effect: Allow
            Action:
              - s3:*
            Resource:
              - !Sub "arn:aws:s3::*:${AthenaOutputBucket}/*"
              - !Sub "arn:aws:s3::*:${AthenaOutputBucket}"
        - Statement:
            Effect: Allow
            Action: s3:GetObject
            Resource:
              - arn:aws:s3::*:membership-dist/*

  QueryLambda:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: !Sub super-mode-calculator-query-lambda-${Stage}
      Description: Queries Athena for article epic views and acquisitions
      Runtime: nodejs14.x
      Handler: lambdas/query/lambda.handler
      MemorySize: 128
      Timeout: 300
      Environment:
        Variables:
          Stage: !Ref Stage
          AthenaOutputBucket: !Ref AthenaOutputBucket
          SchemaName: !Ref SchemaName
      CodeUri:
        Bucket: !Ref DeployBucket
        Key: !Sub ${Stack}/${Stage}/${App}/${App}.zip
      Policies:
        - AWSLambdaBasicExecutionRole
        - AmazonAthenaFullAccess
        - Statement:
            Effect: Allow
            Action:
              - s3:*
            Resource:
              - !Sub "arn:aws:s3::*:${AthenaOutputBucket}/*"
              - !Sub "arn:aws:s3::*:${AthenaOutputBucket}"
        - Statement:
            Effect: Allow
            Action:
              - s3:*
            Resource:
              - !Sub "arn:aws:s3::*:${AcquisitionEventsBucket}/*"
              - !Sub "arn:aws:s3::*:${AcquisitionEventsBucket}"
        - Statement:
            Effect: Allow
            Action: s3:GetObject
            Resource:
              - arn:aws:s3::*:membership-dist/*

  CalculateLambda:
    Type: AWS::Serverless::Function
    Properties:
      FunctionName: !Sub super-mode-calculator-calculate-lambda-${Stage}
      Description: Calculates the list of articles that are in super mode
      Runtime: nodejs14.x
      Handler: lambdas/calculate/lambda.handler
      MemorySize: 128
      Timeout: 300
      Environment:
        Variables:
          Stage: !Ref Stage
      CodeUri:
        Bucket: !Ref DeployBucket
        Key: !Sub ${Stack}/${Stage}/${App}/${App}.zip
      Policies:
        - AWSLambdaBasicExecutionRole
        - AmazonAthenaFullAccess
        - Statement:
            Effect: Allow
            Action:
              - s3:*
            Resource:
              - !Sub "arn:aws:s3::*:${AthenaOutputBucket}/*"
              - !Sub "arn:aws:s3::*:${AthenaOutputBucket}"
        - Statement:
            Effect: Allow
            Action: s3:GetObject
            Resource:
              - arn:aws:s3::*:membership-dist/*
        - Statement:
            Effect: Allow
            Action: dynamodb:*
            Resource:
              - !Sub arn:aws:dynamodb:*:${AWS::AccountId}:table/super-mode-${Stage}
              - !Sub arn:aws:dynamodb:*:${AWS::AccountId}:table/super-mode-${Stage}/index/*

  StateMachineExecutionRole:
    Type: "AWS::IAM::Role"
    Properties:
      AssumeRolePolicyDocument:
        Version: "2012-10-17"
        Statement:
          Effect: "Allow"
          Principal:
            Service: !Sub states.${AWS::Region}.amazonaws.com
          Action: "sts:AssumeRole"
      Path: "/"
      Policies:
      - PolicyName: StatesExecutionPolicy
        PolicyDocument:
          Version: "2012-10-17"
          Statement:
          - Effect: Allow
            Action:
            - lambda:InvokeFunction
            Resource: "*"

  StateMachine:
    Type: "AWS::StepFunctions::StateMachine"
    DependsOn:
      - QueryLambda
      - CalculateLambda
    Properties:
      StateMachineName:
        !Sub ${App}-${Stage}
      DefinitionString:
        !Sub
          - |
            {
              "Comment": "super-mode-calculator",
              "StartAt": "Partition",
              "States": {
                "Partition": {
                  "Type": "Task",
                  "Resource": "${PartitionLambdaArn}",
                  "Next": "Query"
                },
                "Query": {
                  "Type": "Task",
                  "Resource": "${QueryLambdaArn}",
                  "Next": "Calculate",
                  "Retry": [{
                    "ErrorEquals": ["QueryPendingError"],
                    "IntervalSeconds": 10,
                    "MaxAttempts": 5
                  }]
                },
                "Calculate": {
                  "Type": "Task",
                  "Resource": "${CalculateLambdaArn}",
                  "End": true,
                  "Retry": [{
                    "ErrorEquals": ["QueryPendingError"],
                    "IntervalSeconds": 10,
                    "MaxAttempts": 5
                  }]
                }
              }
            }
          -
            PartitionLambdaArn: !GetAtt PartitionLambda.Arn
            QueryLambdaArn: !GetAtt QueryLambda.Arn
            CalculateLambdaArn: !GetAtt CalculateLambda.Arn
      RoleArn: !GetAtt StateMachineExecutionRole.Arn

  TriggerExecutionRole:
    Type: "AWS::IAM::Role"
    Properties:
      AssumeRolePolicyDocument:
        Version: "2012-10-17"
        Statement:
          Effect: Allow
          Principal:
            Service:
            - events.amazonaws.com
          Action: sts:AssumeRole
      Path: "/"
      Policies:
      - PolicyName: StatesExecutionPolicy
        PolicyDocument:
          Version: "2012-10-17"
          Statement:
          - Effect: Allow
            Action:
            - states:StartExecution
            Resource: !Ref StateMachine

  Schedule:
    Type: AWS::Events::Rule
    Properties:
      Name: !Sub super-mode-calculator-schedule-${Stage}
      ScheduleExpression: !Sub cron(${CronExpression})
      State: !Ref ScheduleState
      Targets:
        - Arn: !Ref StateMachine
          RoleArn: !GetAtt TriggerExecutionRole.Arn
          Id: !GetAtt StateMachine.Name

  ExecutionFailureAlarm:
    Type: AWS::CloudWatch::Alarm
    Condition: IsProd
    Properties:
      AlarmActions:
      - !Sub arn:aws:sns:${AWS::Region}:${AWS::AccountId}:alarms-handler-topic-${Stage}
      AlarmName: !Sub ${App}-failed-${Stage}
      AlarmDescription: !Sub Super Mode calculation failed
      MetricName: ExecutionsFailed
      Namespace: AWS/States
      Dimensions:
        - Name: StateMachineArn
          Value: !Ref StateMachine
      ComparisonOperator: GreaterThanOrEqualToThreshold
      Threshold: 1
      Period: 900
      EvaluationPeriods: 1
      Statistic: Sum

