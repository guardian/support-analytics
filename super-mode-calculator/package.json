{
  "name": "super-mode-calculator",
  "version": "1.0.0",
  "license": "MIT",
  "private": true,
  "isAwsLambda": true,
  "cloudformation": "../cfn.yaml",
  "projectName": "Support::super-mode-calculator",
  "buildDir": "./target",
  "riffraffFile": "./riff-raff.yaml",
  "description": "Super mode calculator step functions",
  "scripts": {
    "lint": "eslint '*/**/*.ts'",
    "build": "tsc",
    "local": "Stage=CODE AthenaOutputBucket=gu-support-analytics SchemaName=acquisition AWS_PROFILE=membership ts-node src/local/run.ts"
  },
  "devDependencies": {
    "@guardian/eslint-config-typescript": "^0.7.0",
    "@guardian/prettier": "^2.1.5",
    "@types/node": "^20.8.10",
    "@typescript-eslint/eslint-plugin": "^4.33.0",
    "@typescript-eslint/parser": "^4.33.0",
    "eslint": "^7.32.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^4.2.1",
    "prettier": "^2.8.8",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "aws-sdk": "^2.1488.0",
    "date-fns": "^3.1.0",
    "zod": "^3.22.4"
  }
}
