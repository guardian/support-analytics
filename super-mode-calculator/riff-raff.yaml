stacks: [support]
regions: [eu-west-1]

deployments:
  super-mode-calculator:
    type: aws-lambda
    parameters:
      prefixStack: false
      bucketSsmLookup: true
      fileName: super-mode-calculator.zip
      functionNames:
        - super-mode-calculator-partition-lambda-
        - super-mode-calculator-query-lambda-
        - super-mode-calculator-calculate-lambda-
    dependencies: [super-mode-calculator-cloudformation]
  super-mode-calculator-cloudformation:
    type: cloud-formation
    app: super-mode-calculator
    parameters:
      templatePath: cfn.yaml
