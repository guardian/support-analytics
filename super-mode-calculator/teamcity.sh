#!/bin/bash
set -e

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm

nvm install
nvm use

npm install -g yarn

yarn install

yarn lint

yarn run build

# These also need to be in the RiffRaff package
cp package.json target
cp riff-raff.yaml target

pushd target
# Ensures the RiffRaff package has the node_modules needed to run
yarn install --production
popd

yarn run package
