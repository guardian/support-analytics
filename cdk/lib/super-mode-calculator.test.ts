import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {SuperModeCalculator} from "./super-mode-calculator";

describe('The SuperModeCalculator stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new SuperModeCalculator(app, 'SuperModeCalculator', {
			stack: 'support',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
