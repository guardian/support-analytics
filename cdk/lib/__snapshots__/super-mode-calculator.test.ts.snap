// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SuperModeCalculator stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SuperModeCalculatorC30EED85": {
      "DependsOn": [
        "querylambdaroleDefaultPolicy929B9FB7",
        "querylambdaroleC6C8A94B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/TEST/super-mode/super-mode.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "super-mode",
            "STACK": "support",
            "STAGE": "TEST",
          },
        },
        "FunctionName": "super-mode-TEST",
        "Handler": "lambdas/lambda.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "querylambdaroleC6C8A94B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "super-mode",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-analytics",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SuperModeCalculatorSuperModeCalculatorrate1hour07DE85BC7": {
      "Properties": {
        "ScheduleExpression": "rate(1 hour)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "SuperModeCalculatorC30EED85",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "SuperModeCalculatorSuperModeCalculatorrate1hour0AllowEventRuleSuperModeCalculatorED95104E7D359CFF": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SuperModeCalculatorC30EED85",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "SuperModeCalculatorSuperModeCalculatorrate1hour07DE85BC7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "querylambdaroleC6C8A94B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "super-mode-TEST",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-analytics",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "querylambdaroleDefaultPolicy929B9FB7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/super-mode/TEST/gcp-wif-credentials-config",
                  ],
                ],
              },
            },
            {
              "Action": "dynamodb:Query",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:dynamodb:eu-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/super-mode-calculator-TEST",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:dynamodb:eu-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/super-mode-calculator-TEST/index/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "supermodecalculatortable21F15307",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/TEST/super-mode/super-mode.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/support/super-mode",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/support/super-mode/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "querylambdaroleDefaultPolicy929B9FB7",
        "Roles": [
          {
            "Ref": "querylambdaroleC6C8A94B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "supermodecalculatortable21F15307": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          {
            "AttributeName": "startTimestamp",
            "AttributeType": "S",
          },
          {
            "AttributeName": "endDate",
            "AttributeType": "S",
          },
          {
            "AttributeName": "endTimestamp",
            "AttributeType": "S",
          },
        ],
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "end",
            "KeySchema": [
              {
                "AttributeName": "endDate",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "endTimestamp",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5,
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "startTimestamp",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": false,
        },
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "super-mode-calculator-TEST",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-analytics",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
